src/backend/opencl/kernel/reduce_by_key.hpp

```diff
@@ -484,23 +484,27 @@ int reduce_by_key_first(Array<Tk> &keys_out, Array<To> &vals_out,
                                      sizeof(int), &n_reduced_host);
 
         // reset flags
-        getQueue().enqueueFillBuffer<int>(*needs_another_reduction.get(), 0, 0,
-                                          sizeof(int));
-        getQueue().enqueueFillBuffer<int>(*needs_block_boundary_reduction.get(),
-                                          0, 0, sizeof(int));
-
+        needs_block_boundary_reduction_host = 0;
+        needs_another_reduction_host        = 0;
+
+        getQueue().enqueueWriteBuffer(*needs_another_reduction.get(), CL_FALSE,
```

Comment:
Would be nice to combine these two into an int2 or something. That way you would only need one Write instead of two. Its outside the scope of this PR.